### PR TITLE
Admin can del,show,add user to course

### PR DESCRIPTION
**Pull request by phunghangbk**

### CODE REVIEW

File: app/models/user_course.rb [hunk 1]
```diff
@@ -1,4 +1,6 @@
 class UserCourse < ActiveRecord::Base
   belongs_to :user
   belongs_to :course
+
+  validates :user_id, uniqueness: { scope: :course_id }
```
Comment (leminhtuan2015) [inline app/models/user_course.rb]: Space

File: app/models/user.rb [hunk 2]
```diff
@@ -14,6 +15,10 @@ class User < ActiveRecord::Base
   
   scope :trainee, ->{where.not supervisor: Settings.user.supervisor}
 
+  scope :supervisors, -> { where(supervisor: Settings.user.supervisor) }
```
Comment (leminhtuan2015) [inline app/models/user.rb]: ->{}

File: app/models/user.rb [hunk 3]
```diff
@@ -14,6 +15,10 @@ class User < ActiveRecord::Base
...
+  scope :notsupervisors, -> { where(supervisor: Settings.user.notsupervisor)}
```
Comment (VoTaiTri) [inline app/models/user.rb]: scope ->{}, ->()
```ruby
scope :notsupervisors, ->{where(supervisor: Settings.user.notsupervisor)}
```

File: app/models/user.rb [hunk 4]
```diff
@@ -14,6 +15,10 @@ class User < ActiveRecord::Base
...
+  scope :user_of_course, -> (user_course) { where(:user_courses => {:id => user_course.user_id}).joins(:user_courses)}
```
Comment (VoTaiTri) [inline app/models/user.rb]: try
```ruby
scope :user_of_course, ->user_id{joins(:user_courses).where("user_courses.user_id = ?", user_id)}

File: app/helpers/application_helper.rb [hunk 5]
```diff
@@ -8,12 +8,28 @@ def full_title page_title = ""
     end
   end
 
+  def link_to_add_supervisors label, f, assoc
```
Comment (VoTaiTri) [inline app/helpers/application_helper.rb]: Hàm này có cần thiết ko?
Ko dùng chung hàm link_to_add_fields bên dưới được ah?

File: app/views/admin/courses/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,11 @@
+<% provide :title, t("all_course") %>
+<h1><%= t "all_course" %></h1>
+<ul id="courses">
+<% @courses.each do |course| %>
```
Comment (VoTaiTri) [inline app/views/admin/courses/index.html.erb]: should move to file  app/views/admin/courses/_course.html.erb

File: app/controllers/admin/user_courses_controller.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
+class Admin::UserCoursesController < ApplicationController
```
Comment (VoTaiTri) [inline app/controllers/admin/user_courses_controller.rb]: try to use nested_attributes

File: app/helpers/application_helper.rb (hunk 5 above)
Comment (phunghangbk) [inline app/helpers/application_helper.rb]: Khong dung ham duoi duoc a. Vi link_to_add_fields dung cho nested attributes cua ban Tu.

File: app/controllers/admin/user_courses_controller.rb (hunk 7 above)
Comment (phunghangbk) [inline app/controllers/admin/user_courses_controller.rb]: supervisor co the add mot supervisor khac hoac mot trainee vao mot course nen em phai dung bang nay

Comment (phunghangbk) [inline app/controllers/admin/user_courses_controller.rb]: 1 Supervisor cua mot course co the add them mot supervisor khac hoac mot trainee vao course cua minh nen em khong dung nested_attributes o day a.

File: app/models/user_course.rb [hunk 8]
```diff
@@ -1,4 +1,6 @@
 class UserCourse < ActiveRecord::Base
   belongs_to :user
   belongs_to :course
+
+  validates :user_id, uniqueness: {scope: :course_id}
```
Comment (VoTaiTri) [inline app/models/user_course.rb]: try
```ruby
 validates :user, uniqueness: {scope: :course_id}

File: app/controllers/admin/courses_controller.rb [hunk 9]
```diff
@@ -1,16 +1,25 @@
 class Admin::CoursesController < ApplicationController
-  before_action :logged_in_user, only: [:new, :create, :update]
+  before_action :logged_in_user, only: [:new, :create, :update, :show]
   before_action :is_admin, except: [:index, :show]
 
+  def index
+    @courses = Course.all
```
Comment (dieunb) [inline app/controllers/admin/courses_controller.rb]: pagination

File: app/controllers/admin/user_courses_controller.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+  before_action :logged_in_user, only: [:new, :create, :update]
+  before_action :is_admin, except: [:index, :show]
+  before_action :course, only: [:create, :new]
+  
+  def new
+    @user_course = UserCourse.new
+  end 
+  
+  def create
```
Comment (dieunb) [inline app/controllers/admin/user_courses_controller.rb]: do you need it?

File: app/controllers/admin/user_courses_controller.rb [hunk 11]
```diff
@@ -0,0 +1,29 @@
+class Admin::UserCoursesController < ApplicationController
+  before_action :logged_in_user, only: [:new, :create, :update]
+  before_action :is_admin, except: [:index, :show]
+  before_action :course, only: [:create, :new]
+  
+  def new
```
Comment (dieunb) [inline app/controllers/admin/user_courses_controller.rb]: do you need it?

File: app/controllers/admin/user_courses_controller.rb [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+    byebug
+    if @user_course.save
+      flash[:success] = I18n.t "add_user_succes"
+    else
+      flash[:danger] = I18n.t "add_user_fail"
+    end
+    redirect_to root_path
+  end
+  
+  private
```
Comment (dieunb) [inline app/controllers/admin/user_courses_controller.rb]: do you need it?

File: app/helpers/users_helper.rb [hunk 13]
```diff
@@ -1,5 +1,4 @@
 module UsersHelper
-
```
Comment (dieunb) [inline app/helpers/users_helper.rb]: I don't see any changed things. Please checkout this file

File: app/helpers/application_helper.rb [hunk 14]
```diff
@@ -13,7 +13,7 @@ def link_to_remove_fields label, f
     link = link_to label, "#", onclick: "remove_fields(this)", remote: true
     field + link
   end
-
```
Comment (dieunb) [inline app/helpers/application_helper.rb]: I don't see any changed things. Please checkout this file

File: config/routes.rb [hunk 15]
```diff
@@ -11,7 +11,11 @@
   namespace :admin do
     root  "users#index"
     resources :users
-    resources :courses
+    resources :courses do
+      get "assign_users" => "user_courses#new"
```
Comment (dieunb) [inline config/routes.rb]: it is good if `get "assign_users" => "user_courses#show"`

File: app/controllers/admin/courses_controller.rb [hunk 16]
```diff
@@ -1,16 +1,25 @@
 class Admin::CoursesController < ApplicationController
-  before_action :logged_in_user, only: [:new, :create, :update]
+  before_action :logged_in_user, only: [:new, :create, :update, :show]
```
Comment (VoTaiTri) [inline app/controllers/admin/courses_controller.rb]: should filter by "except"

File: app/models/user.rb [hunk 17]
```diff
@@ -13,6 +14,9 @@ class User < ActiveRecord::Base
   validates :password_confirmation, presence: true
   
   scope :trainee, ->{where.not supervisor: Settings.user.supervisor}
+  scope :supervisors, ->{where(supervisor: Settings.user.supervisor)}
+  scope :notsupervisors, ->{where(supervisor: Settings.user.notsupervisor)}
+  scope :user_of_course, ->user_id{where("id = ?", user_id)}
```
Comment (VoTaiTri) [inline app/models/user.rb]: scope này chỉ là để lấy ra user có id = user_id ???
cái này để làm gì nhỉ?

File: app/controllers/admin/user_courses_controller.rb [hunk 18]
```diff
@@ -0,0 +1,17 @@
+class Admin::UserCoursesController < ApplicationController
+  before_action :logged_in_user, only: [:show]
+  before_action :course, only: [:show]
+
+  def show
+     @user_course = @course.user_courses.find_by course_id: @course.id
```
Comment (dieunb) [inline app/controllers/admin/user_courses_controller.rb]: indent

**Pull request closed without merging**